sqld/src/replication/http.rs

```diff
@@ -0,0 +1,229 @@
+use anyhow::{Context, Result};
+use hyper::server::conn::AddrIncoming;
+use hyper::{Body, Method, Request, Response};
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tower::ServiceBuilder;
+use tower_http::trace::DefaultOnResponse;
+use tower_http::{compression::CompressionLayer, cors};
+use tracing::{Level, Span};
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct FramesRequest {
+    pub next_offset: u64,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct Frames {
+    pub frames: Vec<Frame>,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct Hello {
+    pub generation_id: uuid::Uuid,
+    pub generation_start_index: u64,
+    pub database_id: uuid::Uuid,
+}
+
+pub(crate) async fn run(
+    auth: Arc<Auth>,
+    addr: SocketAddr,
+    logger: Arc<ReplicationLogger>,
+) -> Result<()> {
+    tracing::info!("listening for HTTP requests on {addr}");
+
+    fn trace_request<B>(req: &Request<B>, _span: &Span) {
+        tracing::debug!("got request: {} {}", req.method(), req.uri());
+    }
+    let service = ServiceBuilder::new()
+        .layer(
+            tower_http::trace::TraceLayer::new_for_http()
+                .on_request(trace_request)
+                .on_response(
+                    DefaultOnResponse::new()
+                        .level(Level::DEBUG)
+                        .latency_unit(tower_http::LatencyUnit::Micros),
+                ),
+        )
+        .layer(CompressionLayer::new())
+        .layer(
+            cors::CorsLayer::new()
+                .allow_methods(cors::AllowMethods::any())
+                .allow_headers(cors::Any)
+                .allow_origin(cors::Any),
+        )
+        .service_fn(move |req| {
+            let auth = auth.clone();
+            let logger = logger.clone();
+            handle_request(auth, req, logger)
+        });
+
+    let listener = tokio::net::TcpListener::bind(&addr).await?;
+    let server = hyper::server::Server::builder(AddrIncoming::from_listener(listener)?)
+        .tcp_nodelay(true)
+        .serve(tower::make::Shared::new(service));
+
+    server.await.context("Http server exited with an error")?;
+
+    Ok(())
+}
+
+async fn handle_request(
+    auth: Arc<Auth>,
+    req: Request<Body>,
+    logger: Arc<ReplicationLogger>,
+) -> Result<Response<Body>> {
+    let auth_header = req.headers().get(hyper::header::AUTHORIZATION);
+    let auth = match auth.authenticate_http(auth_header) {
+        Ok(auth) => auth,
+        Err(err) => {
+            return Ok(Response::builder()
+                .status(hyper::StatusCode::UNAUTHORIZED)
+                .body(err.to_string().into())
+                .unwrap());
+        }
+    };
+
+    match (req.method(), req.uri().path()) {
+        (&Method::GET, "/hello") => handle_hello(logger).await,
+        (&Method::POST, "/frames") => handle_query(req, auth, logger).await,
+        _ => Ok(Response::builder().status(404).body(Body::empty()).unwrap()),
+    }
```

Comment:
sure, I'll need to bootstrap myself with Axum first, but if we already have a precedent in the admin API, I'll start by reading that

Reply:
k, done, with one minor quirk: Axum is a little overeager in refusing plaintext HTTP requests when we except a JSON (lack of proper Content-Type), so I decided to be a little more lenient and just parse the string at runtime in case it's a valid json after all.